airtable/airtable_to_mailing_list.py

```diff
@@ -0,0 +1,136 @@
+from apiclient.discovery import build
+from oauth2client.client import SignedJwtAssertionCredentials
+from collections import namedtuple
+
+def get_directory():
+    with open("../config/client_secret.json") as f:
+        secret = json.load(f)
+    credentials = SignedJwtAssertionCredentials(secret["client_email"], secret["private_key"], ["https://www.googleapis.com/auth/admin.directory.group"], sub="samer@directactioneverywhere.com")
+    http = credentials.authorize(httplib2.Http())
+    return build("admin", "directory_v1", http=http)
+
+
+def create_chapter_to_mailing_list(chapters):
+    """Map chapter ids to mailing list emails.
+
+    Takes a list of airtable chapters (that have the form "{ 'id':
+    string, 'fields': { 'Mailing List Email': optional string } }")
+    and returns a dictionary that maps chapter ids to mailing list
+    emails.
+    """
+    chapter_to_mailing_list = {}
+    email_col_name = "Mailing List Email"
+    for c in chapters:
+        if "fields" in c and email_col_name in c["fields"]:
+            email = c["fields"][email_col_name]
+            # Sanity check that the email is probably valid.
+            if "@" not in email:
+                raise Exception("Mailing list email for chapter {} is malformed: {}".format(c, email))
+            c_id = c["id"]
+            chapter_to_mailing_list[c_id] = email
+    return chapter_to_mailing_list
+
+
+# AddMember is a namedtuple for airtable members.
+#  - "airtable_id" is the member's airtable id
+#  - "email" is the member's email
+#  - "chapters_added" is a list of chapter ids where the member has been
+# added to the chapter mailing list
+#  - "chapters_to_add" is a list of chapter ids that represents the
+# chapters that the member hasn't been added to the mailing list for.
+AddMember = namedtuple("AddMember", ["airtable_id", "email", "chapters_added", "chapters_to_add"])
+
+CHAPTER_ID_COL = "chapter_id"
+OPTED_OUT_COL = "Opted Out Of Communication"
+ADDED_TO_MAILING_LIST_COL = "(DO NOT MODIFY) Added To Mailing List"
+
+def get_members_to_add(members):
+    """Filters out members that do not need to be added to a mailing list.
+
+    Filters a list of airtable members that have the form:
+
+    {
+    "id": string,
+    "fields": {
+              "email": optional string,
+              "chapter_id": optional list of strings,
+              "(DO NOT MODIFY) Added To Mailing List": optional list of strings,
+              "Opted Out Of Communication": optional bool
+             }
+    }
+
+    and filters out the members that have 1) opted out of communication,
+    and 2) do not need to be added to a mailing list. Returns a list of
+    AddMembers that need to be added to mailing lists.
+    """
+    members_to_add = []
+    for m in members:
+        if "fields" not in m:
+            continue # m has no fields.
+        fields = m["fields"]
+        # Skip m if they opted out of communication.
+        if fields.get(OPTED_OUT_COL, False):
```

Comment:
I tend to comment the intention of the non-standard path, because they tend to be harder to reason about re: intent, and they're more likely to have bugs.